File: scripts/fetch_news.py
#!/usr/bin/env python3
"""
国土交通省プレスリリース収集・要約スクリプト

国土交通省公式サイトからプレスリリースを取得し、
Claude API で日本語要約を生成して data/news.json に保存する。
"""
import hashlib
import json
import os
import re
import sys
import time
from datetime import datetime, timezone

import anthropic
import requests
from bs4 import BeautifulSoup

MLIT_BASE = "https://www.mlit.go.jp"
MLIT_PRESS_URL = f"{MLIT_BASE}/report/press/index.html"
DATA_FILE = "data/news.json"
MAX_STORED = 60
MAX_NEW_PER_RUN = 10

HEADERS = {
    "User-Agent": "Mozilla/5.0 (compatible; MLITNewsAggregator/1.0; +https://github.com)",
    "Accept-Language": "ja,en;q=0.8",
}

CATEGORY_KEYWORDS = {
    "道路・交通": ["道路", "高速道路", "交通", "渋滞", "ETC", "トンネル", "橋"],
    "鉄道": ["鉄道", "新幹線", "地下鉄", "リニア", "線路", "駅"],
    "航空": ["航空", "空港", "飛行", "ドローン", "UAV"],
    "港湾・海洋": ["港湾", "港", "海運", "船舶", "航路", "海洋"],
    "住宅・建設": ["住宅", "建設", "建築", "マンション", "耐震", "リフォーム", "不動産"],
    "都市・地域": ["都市", "まちづくり", "地域", "コンパクト", "再開発", "区画整理"],
    "河川・治水": ["河川", "治水", "洪水", "ダム", "砂防", "海岸"],
    "気象・防災": ["気象", "防災", "災害", "地震", "津波", "避難"],
    "観光": ["観光", "インバウンド", "旅行", "ツーリズム", "訪日"],
    "自動車": ["自動車", "車両", "EV", "電気自動車", "自動運転", "バス"],
    "水資源": ["水道", "下水", "水資源", "節水"],
    "国土政策": ["国土", "土地", "地価", "公示地価", "不動産登記"],
}


def guess_category(title: str) -> str:
    for cat, keywords in CATEGORY_KEYWORDS.items():
        if any(kw in title for kw in keywords):
            return cat
    return "その他"


def item_id(url: str) -> str:
    return hashlib.md5(url.encode()).hexdigest()[:14]


def fetch(url: str, encoding: str = "utf-8", retries: int = 3) -> str:
    for attempt in range(retries):
        try:
            r = requests.get(url, headers=HEADERS, timeout=30)
            r.encoding = encoding
            r.raise_for_status()
            return r.text
        except Exception as exc:
            if attempt == retries - 1:
                raise
            wait = 2 ** attempt
            print(f"  Retry {attempt+1} after {wait}s ({exc})", file=sys.stderr)
            time.sleep(wait)


def parse_press_list(html: str) -> list[dict]:
    soup = BeautifulSoup(html, "html.parser")
    items = []

    date_re = re.compile(r"\d{4}[年/\-]\d{1,2}[月/\-]\d{1,2}")

    # Strategy 1: <dl> / <dt> + <dd> pairs common in MLIT layout
    for dl in soup.find_all("dl"):
        pairs = list(zip(dl.find_all("dt"), dl.find_all("dd")))
        for dt, dd in pairs:
            link = dd.find("a") or dt.find("a")
            if not link:
                continue
            date_m = date_re.search(dt.get_text())
            _add_item(items, link, date_m.group(0) if date_m else "")

    # Strategy 2: table rows
    if not items:
        for tr in soup.select("table tr"):
            cells = tr.find_all(["td", "th"])
            for cell in cells:
                link = cell.find("a")
                if link:
                    date_m = date_re.search(tr.get_text())
                    _add_item(items, link, date_m.group(0) if date_m else "")

    # Strategy 3: list items with embedded date text
    if not items:
        for li in soup.select("ul li, ol li"):
            link = li.find("a")
            if not link:
                continue
            date_m = date_re.search(li.get_text())
            _add_item(items, link, date_m.group(0) if date_m else "")

    return items


def _add_item(items: list, link, date_str: str):
    title = link.get_text(strip=True)
    href = link.get("href", "")
    if not href or len(title) < 6:
        return
    if href.startswith("http"):
        url = href
    elif href.startswith("/"):
        url = MLIT_BASE + href
    else:
        url = MLIT_BASE + "/" + href
    items.append({"title": title, "date_str": date_str, "url": url})


def parse_date(date_str: str) -> str:
    if not date_str:
        return datetime.now(timezone.utc).strftime("%Y-%m-%d")
    m = re.search(r"(\d{4})[年/\-](\d{1,2})[月/\-](\d{1,2})", date_str)
    if m:
        return f"{m.group(1)}-{int(m.group(2)):02d}-{int(m.group(3)):02d}"
    return datetime.now(timezone.utc).strftime("%Y-%m-%d")


def fetch_article_text(url: str) -> str:
    try:
        html = fetch(url)
        soup = BeautifulSoup(html, "html.parser")
        for tag in soup.select("nav, header, footer, script, style, .nav, .header, .footer, .sidebar"):
            tag.decompose()
        main = soup.select_one("main, #main, .main, article, #content, .content, .body")
        text = (main or soup.body or soup).get_text("\n", strip=True)
        # Collapse blank lines and limit length
        text = re.sub(r"\n{3,}", "\n\n", text)
        return text[:3500]
    except Exception as e:
        print(f"  Could not fetch article: {e}", file=sys.stderr)
        return ""


def summarize(title: str, body: str, client: anthropic.Anthropic) -> str:
    content_part = f"\n\n本文抜粋:\n{body}" if body else ""
    prompt = (
        f"以下の国土交通省プレスリリースを日本語で3〜4文に要約してください。"
        f"施策の目的・内容・効果を中心に簡潔にまとめてください。\n\n"
        f"タイトル: {title}{content_part}"
    )
    try:
        msg = client.messages.create(
            model="claude-haiku-4-5-20251001",
            max_tokens=400,
            messages=[{"role": "user", "content": prompt}],
        )
        return msg.content[0].text.strip()
    except Exception as e:
        print(f"  Claude API error: {e}", file=sys.stderr)
        return title


def load_existing() -> dict:
    if os.path.exists(DATA_FILE):
        with open(DATA_FILE, encoding="utf-8") as f:
            return json.load(f)
    return {"last_updated": "", "items": []}


def save(data: dict):
    os.makedirs(os.path.dirname(DATA_FILE), exist_ok=True)
    with open(DATA_FILE, "w", encoding="utf-8") as f:
        json.dump(data, f, ensure_ascii=False, indent=2)


def main():
    print(f"Fetching MLIT press releases from {MLIT_PRESS_URL}")
    try:
        html = fetch(MLIT_PRESS_URL)
    except Exception as e:
        print(f"ERROR: Could not fetch press list: {e}", file=sys.stderr)
        sys.exit(1)

    press_list = parse_press_list(html)
    print(f"Parsed {len(press_list)} press release entries")

    if not press_list:
        print("WARNING: No entries parsed – page structure may have changed.", file=sys.stderr)

    data = load_existing()
    existing_ids = {item["id"] for item in data.get("items", [])}

    api_key = os.environ.get("ANTHROPIC_API_KEY", "")
    client = anthropic.Anthropic(api_key=api_key) if api_key else None
    if not client:
        print("WARNING: ANTHROPIC_API_KEY not set – skipping summaries.", file=sys.stderr)

    new_items = []
    for entry in press_list:
        uid = item_id(entry["url"])
        if uid in existing_ids:
            continue

        print(f"  + {entry['title'][:70]}")

        article_body = ""
        if client:
            article_body = fetch_article_text(entry["url"])
            time.sleep(0.5)

        summary = summarize(entry["title"], article_body, client) if client else entry["title"]

        new_items.append({
            "id": uid,
            "date": parse_date(entry["date_str"]),
            "title": entry["title"],
            "url": entry["url"],
            "summary": summary,
            "category": guess_category(entry["title"]),
        })

        if len(new_items) >= MAX_NEW_PER_RUN:
            break

    print(f"Added {len(new_items)} new items")

    merged = new_items + data.get("items", [])
    data["items"] = merged[:MAX_STORED]
    data["last_updated"] = datetime.now(timezone.utc).isoformat()

    save(data)
    print(f"Saved {len(data['items'])} total items to {DATA_FILE}")


if __name__ == "__main__":
    main()

File: scripts/fetch-news.mjs
#!/usr/bin/env node
// 国土交通省 (MLIT) のプレスリリースを取得し、Anthropic API で要約して
// data/news.json に保存するスクリプト。
//
// 必要な環境変数:
//   ANTHROPIC_API_KEY  - Anthropic API キー (任意。未指定時は要約をスキップ)
//
// 実行: node scripts/fetch-news.mjs

import { writeFile, mkdir, readFile } from "node:fs/promises";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const REPO_ROOT = join(__dirname, "..");
const DATA_DIR = join(REPO_ROOT, "data");
const OUTPUT_FILE = join(DATA_DIR, "news.json");

const MLIT_PRESS_INDEX = "https://www.mlit.go.jp/report/press/index.html";
const MLIT_BASE = "https://www.mlit.go.jp";

const MAX_ITEMS = 30;
const SUMMARIZE_TOP_N = 15;
const ANTHROPIC_MODEL = "claude-haiku-4-5-20251001";

const UA =
  "Mozilla/5.0 (compatible; MLIT-News-Watcher/1.0; +https://github.com/yusa-san-land/claude-test)";

async function fetchText(url) {
  const res = await fetch(url, {
    headers: { "User-Agent": UA, Accept: "text/html,application/xhtml+xml" },
    redirect: "follow",
  });
  if (!res.ok) {
    throw new Error(`Failed to fetch ${url}: ${res.status} ${res.statusText}`);
  }
  const buf = await res.arrayBuffer();
  // MLIT のページは UTF-8 で配信されているが、念のため明示
  return new TextDecoder("utf-8").decode(buf);
}

function decodeHtmlEntities(s) {
  return s
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&nbsp;/g, " ");
}

function stripTags(s) {
  return decodeHtmlEntities(s.replace(/<[^>]+>/g, "")).replace(/\s+/g, " ").trim();
}

function absoluteUrl(href) {
  if (!href) return null;
  if (/^https?:\/\//i.test(href)) return href;
  if (href.startsWith("//")) return "https:" + href;
  if (href.startsWith("/")) return MLIT_BASE + href;
  return new URL(href, MLIT_PRESS_INDEX).toString();
}

// MLIT のプレスリリース一覧から (date, title, url) を抽出する。
// ページ構造は時期によって若干変わるため、複数のパターンを試す。
function parsePressReleases(html) {
  const items = [];
  const seen = new Set();

  // パターン: <a href="...">タイトル</a> の近傍に日付 (YYYY年M月D日 / YYYY/MM/DD など)
  const linkRe =
    /<a\s+[^>]*href=["']([^"']+)["'][^>]*>([\s\S]*?)<\/a>/gi;
  const dateRe =
    /(20\d{2})\s*[年\/\-\.]\s*(\d{1,2})\s*[月\/\-\.]\s*(\d{1,2})/;

  // 一覧の各行はだいたい "日付 ... <a>タイトル</a>" の形をしている。
  // 行単位に分割して走査する。
  const lines = html.split(/<\/?(?:li|tr|dt|dd|p|br\s*\/?)>/i);
  for (const line of lines) {
    const dateMatch = line.match(dateRe);
    if (!dateMatch) continue;
    const [, y, m, d] = dateMatch;
    const date = `${y}-${String(m).padStart(2, "0")}-${String(d).padStart(2, "0")}`;

    linkRe.lastIndex = 0;
    let lm;
    while ((lm = linkRe.exec(line)) !== null) {
      const href = absoluteUrl(lm[1]);
      const title = stripTags(lm[2]);
      if (!href || !title) continue;
      if (title.length < 4) continue;
      // ナビゲーションリンクなどを除外
      if (/^(トップ|ホーム|前へ|次へ|一覧|サイトマップ|english)$/i.test(title)) continue;
      // プレスリリース系のリンクに絞る
      if (!/mlit\.go\.jp/.test(href)) continue;
      if (/index\.html?$/i.test(href)) continue;

      const key = href + "|" + title;
      if (seen.has(key)) continue;
      seen.add(key);

      items.push({ date, title, url: href });
      if (items.length >= MAX_ITEMS) break;
    }
    if (items.length >= MAX_ITEMS) break;
  }

  return items;
}

async function summarizeWithAnthropic(apiKey, items) {
  if (!apiKey) return items;

  const toSummarize = items.slice(0, SUMMARIZE_TOP_N);
  const numbered = toSummarize
    .map((it, i) => `${i + 1}. (${it.date}) ${it.title}`)
    .join("\n");

  const prompt = `あなたは国土交通省のプレスリリースを分かりやすく解説するアナリストです。
以下は最近の国土交通省プレスリリースのタイトル一覧です。
それぞれについて、以下を出力してください:

- summary: タイトルから推測できる内容を、一般読者向けに 60〜120 字の日本語で要約
- category: 次のいずれかから最も近いもの ["道路・交通", "鉄道", "航空", "海事・港湾", "住宅・建築", "都市・地域", "防災・気象", "観光", "物流", "DX・データ", "国際", "その他"]
- importance: "high" | "normal"  (国民生活・経済への影響が大きそうなら high)

必ず JSON のみを返してください。形式:
{
  "items": [
    { "index": 1, "summary": "...", "category": "...", "importance": "..." },
    ...
  ]
}

タイトル一覧:
${numbered}`;

  const res = await fetch("https://api.anthropic.com/v1/messages", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "x-api-key": apiKey,
      "anthropic-version": "2023-06-01",
    },
    body: JSON.stringify({
      model: ANTHROPIC_MODEL,
      max_tokens: 4000,
      messages: [{ role: "user", content: prompt }],
    }),
  });

  if (!res.ok) {
    const text = await res.text();
    console.error(`Anthropic API error ${res.status}: ${text}`);
    return items;
  }

  const json = await res.json();
  const text = json?.content?.[0]?.text ?? "";
  const match = text.match(/\{[\s\S]*\}/);
  if (!match) {
    console.error("Failed to parse Anthropic response as JSON:", text);
    return items;
  }

  let parsed;
  try {
    parsed = JSON.parse(match[0]);
  } catch (e) {
    console.error("JSON parse error:", e.message);
    return items;
  }

  const byIndex = new Map();
  for (const r of parsed.items ?? []) {
    if (typeof r.index === "number") byIndex.set(r.index, r);
  }

  return items.map((it, i) => {
    const r = byIndex.get(i + 1);
    if (!r) return it;
    return {
      ...it,
      summary: r.summary,
      category: r.category,
      importance: r.importance,
    };
  });
}

async function main() {
  console.log(`Fetching ${MLIT_PRESS_INDEX} ...`);
  const html = await fetchText(MLIT_PRESS_INDEX);
  const parsed = parsePressReleases(html);
  console.log(`Parsed ${parsed.length} press release entries.`);

  if (parsed.length === 0) {
    // 既存ファイルがある場合は壊さないため、上書きしない
    console.warn("No items parsed; keeping existing data/news.json if present.");
    try {
      await readFile(OUTPUT_FILE, "utf-8");
      console.warn("Existing file kept. Exiting without changes.");
      return;
    } catch {
      // 既存ファイルが無い場合は空のデータを書き込む
    }
  }

  const apiKey = process.env.ANTHROPIC_API_KEY;
  if (!apiKey) {
    console.warn("ANTHROPIC_API_KEY not set; skipping summarization.");
  }

  const enriched = await summarizeWithAnthropic(apiKey, parsed);

  const output = {
    source: MLIT_PRESS_INDEX,
    fetched_at: new Date().toISOString(),
    summarized: Boolean(apiKey),
    count: enriched.length,
    items: enriched,
  };

  await mkdir(DATA_DIR, { recursive: true });
  await writeFile(OUTPUT_FILE, JSON.stringify(output, null, 2) + "\n", "utf-8");
  console.log(`Wrote ${OUTPUT_FILE} (${enriched.length} items).`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
